consensus/core/src/transaction_certifier.rs

```diff
@@ -134,33 +100,46 @@ impl TransactionCertifier {
                 .scan_blocks_by_author(authority_index, recovery_start_round)
                 .unwrap();
             info!(
-                "Recovering {} blocks for authority {}",
+                "Recovering and voting on {} blocks for authority {} {}",
                 blocks.len(),
+                authority_index,
                 context.committee.authority(authority_index).hostname
             );
-            let voted_blocks = blocks
-                .into_iter()
-                .map(|b| {
-                    if b.round() <= dag_state_gc_round || dag_state.is_hard_linked(&b.reference()) {
-                        // Own votes are unnecessary for blocks already included in own blocks,
-                        // or outside of local DAG GC bound.
-                        (b, vec![])
-                    } else {
-                        // Own votes are needed for blocks not yet included in own blocks. They will be
-                        // added to proposed blocks together.
-                        let reject_transaction_votes =
-                            block_verifier.vote(&b).unwrap_or_else(|e| {
-                                panic!("Failed to vote on block during recovery: {}", e)
-                            });
-                        (b, reject_transaction_votes)
-                    }
-                })
-                .collect::<Vec<_>>();
-            let certified_blocks = certifier_state.add_voted_blocks(voted_blocks);
-            self.send_certified_blocks(certified_blocks);
+            self.recover_and_vote_on_blocks(blocks);
         }
     }
 
+    /// Recovers and votes on the given blocks.
+    ///
+    /// The recovery processes peers' reject votes contained in the input blocks.
+    /// and potentially votes on the input blocks as well..
+    ///
+    /// Because own votes on blocks are not stored, so blocks are voted on if they can be included in
+    /// a future proposed block.
+    pub(crate) fn recover_and_vote_on_blocks(&self, blocks: Vec<VerifiedBlock>) {
+        let dag_state = self.dag_state.read();
+        let voted_blocks = blocks
+            .into_iter()
+            .map(|b| {
+                if b.round() <= dag_state.gc_round() || dag_state.is_hard_linked(&b.reference()) {
+                    // Own votes are unnecessary for blocks already included in own blocks,
+                    // or outside of local DAG GC bound.
+                    (b, vec![])
+                } else {
+                    // Own votes are needed for blocks not yet included in own blocks. They will be
+                    // added to proposed blocks together.
+                    let reject_transaction_votes =
```

Comment:
Can you walk me through why these blocks are guaranteed to be added to an own proposed block after this point? Can't ancestor selection still exclude them? If so would it be wrong to add voted blocks here